CMS-Enterprise/mint-app
# [Mint-3129]/Dataloaders new version refactor

# MINT-3129

## Description

Since data loaders were first implemented in MINT, there have been some notable improvements in libraries. We initially looked to implement dataloadgen as a replacement to graph gophers, but instead decided to implement a more recent version of graph gophers which implements generics. The generic structure gives us some safety that was lacking, and also allows us to handle caching discretely (which is not allows in dataloadgen).

Also, this PR looks to abstract some of the repetitive patterns we have had to implement in the dataloaders, and instead utilize some generic structs to link methods together. 

Notable Changes
1.  Loaders grouped and exported by type
    a. See example for ModelPlan, Plan Basics, and Operational Solution
    b. Especially note how Operational Solution has two loader configs.
2. Note, loaders are no longer receiver methods, so they can be associated with a configuration
3. Loaders from context now returns a bool if the loader exists on the context or not
4. Dataloader new methods aren't being exported out of the package, the loader config from the model is now exported to support a semantic method of referencing a data loader,
    a. EX
       i. `loaders.OperationalSolutions.ByID.Load(ctx, id)`
       ii. `	return loaders.OperationalSolutions.AndPossibleByOperationalNeedID.Load(ctx,
		storage.SolutionAndPossibleKey{
			OperationalNeedID: operationalNeedID,
			IncludeNotNeeded:  includeNotNeeded,
		})`
5. Store methods for loaders are type safe thanks to the new generic types. If data is being serialized to JSON, that happens internally in the store method. See [OperationalSolutionExample](https://github.com/CMSgov/mint-app/blob/8f7a51547f1079f23d5278c42ba6efe323925ccf/pkg/storage/operational_solutionStore.go#L61)
6. Simple use cases can pass arrays as pq.Array. See [ModelPlan Example](https://github.com/CMSgov/mint-app/blob/5e55402c05cbb30ae2c6d8b487070319c1751fc7/pkg/storage/model_planStore.go#L27)
		


## How to test this change

1. Verify tests pass.
3. Hit dataloaders work through postman if you like
4. Mostly, read the code and make sure it makes sense to you

## PR Author Checklist




## PR Reviewer Guidelines

- It's best to pull the branch locally and test it, rather than just looking at the code online!
- When approving a PR, provide a reason _why_ you're approving it
  - e.g. "Approving because I tested it locally and all functionality works as expected"
  - e.g. "Approving because the change is simple and matches the Figma design"
- Don't be afraid to leave comments or ask questions, especially if you don't understand why something was done! (This is often a great time to suggest code comments or documentation updates)
- Check that all code is adequately covered by tests - if it isn't feel free to suggest the addition of tests.
